src/kernel/pgtbl.c

```diff
@@ -4,184 +4,145 @@
-		return -ECASFAIL;
-	}
-	/* Return the pte ptr, so that we can release the page if the
-	 * kobj activation failed later. */
-	*pte_ret = (unsigned long *)pte;
+int
+pgtbl_deactivate(struct captbl *t, struct cap_captbl *dest_ct_cap, unsigned long capin,
+                 livenessid_t lid, capid_t pgtbl_cap, capid_t cosframe_addr, const int root)
+{
+	return chal_pgtbl_deactivate(t, dest_ct_cap, capin, lid, pgtbl_cap, cosframe_addr, root);
+}
 
-	return 0;
+/**
+ * this works on both kmem and regular user memory: the retypetbl_ref
+ * works on both.
+ */
+int
+pgtbl_mapping_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
+{
+	return chal_pgtbl_mapping_add(pt, addr, page, flags);
 }
 
-/* Return 1 if quiescent past since input timestamp. 0 if not. */
+/**
+ * This function is only used by the booting code to add cos frames to
+ * the pgtbl. It ignores the retype tbl (as we are adding untyped
+ * frames). 
+ */
 int
-tlb_quiescence_check(u64_t timestamp)
+pgtbl_cosframe_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
 {
-	int i, quiescent = 1;
-
-	/* Did timer interrupt (which does tlb flush
-	 * periodically) happen after unmap? The periodic
-	 * flush happens on all cpus, thus only need to check
-	 * the time stamp of the current core for that case
-	 * (assuming consistent time stamp counters). */
-	if (timestamp > tlb_quiescence[get_cpuid()].last_periodic_flush) {
-		/* If no periodic flush done yet, did the
-		 * mandatory flush happen on all cores? */
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			if (timestamp > tlb_quiescence[i].last_mandatory_flush) {
-				/* no go */
-				quiescent = 0;
-				break;
-			}
-		}
-	}
-	if (quiescent == 0) {
-		printk("from cpu %d, t %llu: cpu %d last mandatory flush: %llu\n", get_cpuid(), timestamp, i,
-		       tlb_quiescence[i].last_mandatory_flush);
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
-		}
-	}
-
-	return quiescent;
+	return chal_pgtbl_cosframe_add(pt, addr, page, flags);
 }
 
+/* This function updates flags of an existing mapping. */
 int
-cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
+pgtbl_mapping_mod(pgtbl_t pt, u32_t addr, u32_t flags, u32_t *prevflags)
 {
-	unsigned long *    pte, cosframe, orig_v;
-	struct cap_header *dest_pt_h;
-	u32_t              flags;
-	int                ret;
+	return chal_pgtbl_mapping_mod(pt, addr, flags, prevflags);
+}
 
-	if (unlikely(pt->lvl || (pt->refcnt_flags & CAP_MEM_FROZEN_FLAG))) return -EINVAL;
+/**
+ * When we remove a mapping, we need to link the vas to a liv_id,
+ * which tracks quiescence for us.
+ */
+int
+pgtbl_mapping_del(pgtbl_t pt, u32_t addr, u32_t liv_id)
+{
+	return chal_pgtbl_mapping_del(pt, addr, liv_id);
+}
 
-	dest_pt_h = captbl_lkup(ct, dest_pt);
-	if (dest_pt_h->type != CAP_PGTBL) return -EINVAL;
-	if (((struct cap_pgtbl *)dest_pt_h)->lvl) return -EINVAL;
+/* 
```

Comment:
I'm sorry to be nitpicking so much but the comment style should really be uniform. 
I'm talking about the use of `/*` vs `/**`.

Reply:
```
/**
 * this is a comment that should appear in high-level documentation for the system, perhaps auto-generated by something like doxygen
 */
```

```
/*
 * This is a normal comment that
 * spans multiple lines.
 */
```

Given this, I think that using `/*` instead of `/**` is most appropriate for most of these.

Reply:
I see.